File: 2021/07/26/sorted_array_to_bst.py
"""
https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
Author: Andrew Jarombek
Date: 7/26/2021
"""

# You are still gone.  If you are happier out there without me I’m happy for you, even though I miss you.

from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def sorted_array_to_bst(nums: List[int]) -> Optional[TreeNode]:
    if len(nums) == 0:
        return None
    elif len(nums) == 1:
        return TreeNode(nums[0])
    else:
        middle = len(nums) // 2
        return TreeNode(
            nums[middle],
            sorted_array_to_bst(nums[0:middle]),
            sorted_array_to_bst(nums[middle + 1:len(nums)])
        )


if __name__ == '__main__':
    bst = sorted_array_to_bst([-10, -3, 0, 5, 9])
    assert bst.val == 0
    assert bst.left.val == -3
    assert bst.right.val == 9
    assert bst.left.left.val == -10
    assert bst.right.left.val == 5

File: 2022/01/29/c/sortedArrayToBST.c
// Program: https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
// Author: Andrew Jarombek
// Date: 1/29/2022

#include <stddef.h>
#include <stdlib.h>
#include <assert.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct TreeNode* createTree(int* nums, struct TreeNode* node, int start, int end) {
    if (start == end) {
        node->val = nums[start];
        node->left = NULL;
        node->right = NULL;
        return node;
    } else {
        int mid = (start + end) / 2;
        node->val = nums[mid];
        node->left = NULL;
        node->right = NULL;

        if (start != mid) {
            struct TreeNode* left = malloc(sizeof(struct TreeNode));
            node->left = left;
            createTree(nums, left, start, mid - 1);
        }

        if (end != mid) {
            struct TreeNode* right = malloc(sizeof(struct TreeNode));
            node->right = right;
            createTree(nums, right, mid + 1, end);
        }
    }

    return node;
}

struct TreeNode* sortedArrayToBST(int* nums, int numsSize){
    struct TreeNode* node = malloc(sizeof(struct TreeNode));
    createTree(nums, node, 0, numsSize - 1);
    return node;
}

int main() {
    int* nums = malloc(sizeof(int) * 5);
    nums[0] = 1;
    nums[1] = 2;
    nums[3] = 3;

    struct TreeNode* node = sortedArrayToBST(nums, 3);

    free(nums);
    free(node);
}
